include/toypp/sharedptr.hpp

```diff
@@ -0,0 +1,224 @@
+
+  struct Shared {
+    std::atomic<std::size_t> ptr_refcount = 1;
+    std::atomic<std::size_t> shared_refcount = 1;
+    std::function<void(value_type*)> deleter = k_default_deleter;
+  };
+
+  inline static Shared s_null_shared{.ptr_refcount = 0, .shared_refcount = 1, .deleter = k_null_deleter};
+
+  value_type* _ptr;
+  Shared* _shared;
+
+  friend class WeakPtr<T>;
+  friend class WeakPtr<value_type>;
+
+  void destroy() {
+    if (!_shared) {
+      return;
+    }
+    if (_shared->ptr_refcount.fetch_sub(1) == 1) {
+      _shared->deleter(_ptr);
+    }
+    if (_shared->shared_refcount.fetch_sub(1) == 1) {
+      delete _shared;
+    }
+  }
+
+  explicit SharedPtr(value_type* ptr, Shared* shared) noexcept : _ptr(ptr), _shared(shared) {}
+
+ public:
+  SharedPtr() noexcept : _ptr(nullptr), _shared(&s_null_shared) {
+    _shared->ptr_refcount.fetch_add(1);
+    _shared->shared_refcount.fetch_add(1);
+  }
+  SharedPtr(value_type* ptr) : _ptr(ptr), _shared(new Shared{}) {}
+
+  template <typename F>
+  SharedPtr(value_type* ptr, F&& deleter) : _ptr(ptr), _shared(new Shared{.deleter = std::forward<F>(deleter)}) {}
+
+  SharedPtr(const SharedPtr& other) : _ptr(other._ptr), _shared(other._shared) {
+    other._shared->ptr_refcount.fetch_add(1);
+    other._shared->shared_refcount.fetch_add(1);
+  }
+  SharedPtr(SharedPtr&& other) noexcept
+    : _ptr(std::exchange(other._ptr, nullptr))
+    , _shared(std::exchange(other._shared, nullptr))
+  {}
+
+  SharedPtr& operator=(const SharedPtr& other) {
+    if (this == &other) {
+      return *this;
+    }
+    destroy();
+    _ptr = other._ptr;
+    _shared = other._shared;
+    other._shared->ptr_refcount.fetch_add(1);
+    other._shared->shared_refcount.fetch_add(1);
+    return *this;
+  }
+  SharedPtr& operator=(SharedPtr&& other) noexcept {
+    std::swap(_ptr, other._ptr);
+    std::swap(_shared, other._shared);
+    return *this;
+  }
+
+  ~SharedPtr() { destroy(); }
+
+  value_type* get() { return _ptr; }
+  const value_type* get() const { return _ptr; }
+
+  std::size_t use_count() const { return _shared->ptr_refcount.load(); }
+
+  void reset() { reset(nullptr, k_null_deleter); }
+  void reset(value_type* ptr) { reset(ptr, k_default_deleter); }
+
+  template <typename F>
+  void reset(value_type* ptr, F&& deleter) {
+    destroy();
+
+    // to make sure if `new Shared` fails, the object will be in a valid state.
+    _ptr = nullptr;
+    _shared = &s_null_shared;
+
+    auto shared = new Shared{.deleter = std::forward<F>(deleter)};
+    _ptr = ptr;
+    _shared = shared;
+  }
+
+  bool operator==(const SharedPtr& other) const { return _ptr == other._ptr; }
+  bool operator!=(const SharedPtr& other) const { return _ptr != other._ptr; }
+
+  value_type* operator->() { return _ptr; }
+  const value_type* operator->() const { return _ptr; }
+
+  value_type& operator*() & { return *_ptr; }
+  const value_type& operator*() const& { return *_ptr; }
```

Comment:
I don't think there should be a `const` overload of `operator*` or `operator->`.  I don't think the `const`-ness of a `SharedPtr` should propagate to the referenced object, since `SharedPtr` is more of a pointer type that refers to its object than a container type that owns its object.  This opinion is not universally shared; some others might feel otherwise.